traffic_ops/traffic_ops_golang/deliveryservice/deleteoldcerts.go

```diff
@@ -0,0 +1,197 @@
+	dbCtx, cancelTx := context.WithTimeout(context.Background(), dbTimeout)
+	tx, err := db.BeginTx(dbCtx, nil)
+	if err != nil {
+		log.Errorln("Old Cert Deleter Job: beginning tx: " + err.Error())
+		return
+	}
+	defer cancelTx()
+	txCommit := false
+	defer dbhelpers.CommitIf(tx, &txCommit)
+	if err := deleteOldDSCerts(tx, riakOpts, riakPort, cdn); err != nil {
+		log.Errorln("deleting old DS certificates: " + err.Error())
+		return
+	}
+	txCommit = true
+}
+
+var oldCertDeleters = OldCertDeleters{D: map[tc.CDNName]*OldCertDeleter{}}
+
+type OldCertDeleters struct {
+	D map[tc.CDNName]*OldCertDeleter
+	M sync.Mutex
+}
+
+// startOldCertDeleter tells the old cert deleter goroutine to start another delete job, creating the goroutine if it doesn't exist.
+func startOldCertDeleter(db *sql.DB, tx *sql.Tx, dbTimeout time.Duration, riakOpts *riak.AuthOptions, riakPort *uint, cdn tc.CDNName) {
+	oldCertDeleter := getOrCreateOldCertDeleter(cdn)
+	oldCertDeleter.Once.Do(func() {
+		go doOldCertDeleter(oldCertDeleter.Start, oldCertDeleter.Die, db, dbTimeout, riakOpts, riakPort, cdn)
+	})
+
+	select {
+	case oldCertDeleter.Start <- struct{}{}:
+	default:
+	}
+}
+
+func getOrCreateOldCertDeleter(cdn tc.CDNName) *OldCertDeleter {
+	oldCertDeleters.M.Lock()
+	defer oldCertDeleters.M.Unlock()
+	oldCertDeleter, ok := oldCertDeleters.D[cdn]
+	if !ok {
+		oldCertDeleter = newOldCertDeleter()
+		oldCertDeleters.D[cdn] = oldCertDeleter
+	}
+	return oldCertDeleter
+}
+
+// cleanupOldCertDeleters stops all cert deleter goroutines for CDNs which have been deleted in the database.
+// Any error is logged, but not returned.
+// This is designed to be called when starting a new cert deleter job, to clean up any old cert deleters from deleted CDNs.
+// This should only be called immediately when snapshotting, and immediately after startOldCertDeleter, because otherwise a cert deleter may be removed before it can delete old certs for a given CDN.
+func cleanupOldCertDeleters(tx *sql.Tx) {
+	cdns, err := dbhelpers.GetCDNs(tx) // (map[tc.CDNName]struct{}, error) {
+	if err != nil {
+		log.Errorln("StopOldCertDeleters: getting cdns: " + err.Error())
+		return
+	}
+
+	oldCertDeleters.M.Lock()
+	defer oldCertDeleters.M.Unlock()
+
+	for cdn, oldCertDeleter := range oldCertDeleters.D {
+		if _, ok := cdns[cdn]; ok {
+			continue
+		}
+		delete(oldCertDeleters.D, cdn)
+		select {
+		case oldCertDeleter.Die <- struct{}{}:
+		default:
+		}
+	}
+}
+
+type OldCertDeleter struct {
+	Start chan struct{}
+	Die   chan struct{}
+	Once  sync.Once
+}
+
+func newOldCertDeleter() *OldCertDeleter {
+	return &OldCertDeleter{
+		Start: make(chan struct{}, 1),
+		Die:   make(chan struct{}, 1),
+	}
+}
+
+func doOldCertDeleter(do chan struct{}, die chan struct{}, db *sql.DB, dbTimeout time.Duration, riakOpts *riak.AuthOptions, riakPort *uint, cdn tc.CDNName) {
+	for {
+		select {
+		case <-do:
+			deleteOldDSCertsDB(db, dbTimeout, riakOpts, riakPort, cdn)
```

Comment:
could this be simplified into the following?
```
for _ := range do {
    ...
}
```
Then this `select` wouldn't be necessary since we would no longer need the `die chan`, and `cleanupOldCertDeleters` could just `close(oldCertDeleter.Start)` instead of `select`ing as well.

This might also prevent the case where there is an outstanding "delete old CDN certs" job in the `do` channel but the CDN was deleted at the same time.

Reply:
Go channels can only be safely closed by a single writer goroutine. If `close` is called on an already closed chan, it panics. Here, any number of client requests goroutines could be executing the cleanup. This would introduce a race condition, if multiple requests happened to be in the block at once.

It's possible to make it safe, by either ensuring the `OldCertDeleters` mutex is always locked when the channel is used, or adding a mutex for each `OldCertDeleter` chan.

There would also be another race when writing to the chan, if another request goroutine closed it. So, either another lock would have to be introduced in the start, or code would have to be added to check if the chan was closed before writing to it. 

Go also has no way to check if a chan is closed without reading from it, so if either the writer or closer checked for a close, it would have to also write back what it just read.

So, it's possible to use `range` and `close`, but it wouldn't be simpler. Further, the map mutex is inevitable (unless a `sync.Map` were used), but mutexing channels is very unexpected. Because `chan` is defined to be safe for multiple goroutines, it seems likely someone might come along in the future, not fully understand the race conditions because "chan is threadsafe", and remove mutexes to simplify things, introducing races.

>This might also prevent the case where there is an outstanding "delete old CDN certs" job in the `do` channel but the CDN was deleted at the same time.

I'm not sure what you mean exactly, but we want the final delete job to go through, even if the channel is closed. For example, if someone snapshots and deletes the cdn at the same time.

Hence, a `range`/`close` change would also have to account for that, and be careful not to introduce races around that, such that a final delete job didn't get executed.

Reply:
Ah, I didn't realize closing a closed channel actually panics (thanks, golang).

I was thinking that this block:
```
	for {
		select {
		case <-do:
			deleteOldDSCertsDB(db, dbTimeout, riakOpts, riakPort, cdn)
		case <-die:
			return
		}
	}
```
would potentially read from the `die` channel when there is an outstanding job in the `do` channel, which would cause the goroutine to exit without handling that final delete job. This is because it reads from a random channel if multiple are ready.

Can we just `defer` a `recover()` in case the channel has already been closed by another goroutine?

Reply:
>Can we just `defer` a `recover()` in case the channel has already been closed by another goroutine?

We could, but that's very un-idiomatic. Go chans panicing on double-close is an intentional decision by the Go authors to suggest you should rethink your design, if you need that. I'd really prefer not to write code with a known race, that violates good design principles, handled by `recover`. A `die` chan is clean and obvious what it's doing, and I think is the common and accepted way of doing this in Go.

>would potentially read from the `die` channel when there is an outstanding job in the `do` channel, which would cause the goroutine to exit without handling that final delete job.

You're right. Though I think it would have to be two writes, a second write while a job is already executing, concurrently with a CDN delete and snapshot. The "perfect" way to handle that is:

```
	for {
		select {
		case <-do:
			deleteOldDSCertsDB(db, dbTimeout, riakOpts, riakPort, cdn)
		case <-die:
			select {
			case <-do:
				deleteOldDSCertsDB(db, dbTimeout, riakOpts, riakPort, cdn)
			default:
			}
			return
		}
	}
```

You can see the Monitor doing this in places. I'd decided it wasn't worth the readability/complexity for the odds of that triple race happening. But in retrospect, I can agree, even the unlikeliest races cause outages years later when you least expect. I'll change it.

Reply:
Done.

Reply:
Yeah I was thinking of the Python way of "do it now and ask forgiveness later". I think your nested selects get the job done too and are not that complicated to read IMO.